### ISSUE TITLE
2015.8.5 explodes if salt-call is mistakenly not run as root

### ISSUE THREAD

**Issue by alangarf** [saltstack/salt#30925]
Hey all, this issue with the pki dir has been around for a while (#29701) and the solution was to add a `mkdir` to the packages. This now is presenting itself again but in an unexpected way.

Any how, if you run `salt-call test.version` you will be presented with:
```
[WARNING ] Failed to open log file, do you have permission to write to /var/log/salt/minion?
[WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate.
[ERROR   ] Future exception was never retrieved: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result)
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 500, in eval_master
    pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/transport/client.py", line 154, in factory
    return salt.transport.zeromq.AsyncZeroMQPubChannel(opts, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 253, in __init__
    self.auth = salt.crypt.AsyncAuth(self.opts, io_loop=self.io_loop)
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 324, in __new__
    new_auth.__singleton_init__(opts, io_loop=io_loop)
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 362, in __singleton_init__
    self.get_keys()
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 576, in get_keys
    salt.utils.verify.check_path_traversal(self.opts['pki_dir'], user)
  File "/usr/lib/python2.7/dist-packages/salt/utils/verify.py", line 400, in check_path_traversal
    raise SaltClientError(msg)
SaltClientError: Could not access /etc/salt/pki/minion. Try running as user root.
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7f68e5b93e90>> ignored
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7f68e5b93e90>> ignored
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7f68e5b93e90>> ignored
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7f68e5b93e90>> ignored
[LOOP FOREVER TILL KILLED]
```

... but when correctly run it as `sudo salt-call test.version` you get:

```
local:
    2015.8.5
```

I know I was using the wrong credentials, but surely that shouldn't descend into a massive loop of output.

**Comment by jmo-learnosity** [comment]
I am able to reproduce this as well. It completely bombs if you run it without sudo with any command (not just `test.version`, but seems to execute without exploding if you give the `--local` flag (on Ubuntu 14.04, at least):
```
$ salt-call --local test.version
[WARNING ] Failed to open log file, do you have permission to write to /var/log/salt/minion?
[WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate.
Cannot write to process directory. Do you have permissions to write to /var/cache/salt/minion/proc/20160205083618949955 ?
local:
    2015.8.5
```

**Comment by ssgward** [comment]
@alangarf - is this the same issue: #29820? Seems to be the exact opposite from what you're describing, though.

**Comment by jmo-learnosity** [comment]
@ssgward Seems to me that is what he is describing. I'm recreating this on a machine with both salt-master and salt-minion installed. When I run `salt-call test.version -l debug`, it gives me this:
```
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/local.conf
[DEBUG   ] Reading configuration from /etc/salt/minion.d/local.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: master.local
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Failed to open log file, do you have permission to write to /var/log/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/local.conf
[DEBUG   ] Reading configuration from /etc/salt/minion.d/local.conf
[DEBUG   ] Please install 'virt-what' to improve results of the 'virtual' grain
[WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'master.local', 'tcp://127.0.0.1:4506
[ERROR   ] Future exception was never retrieved: Traceback (most recent call last
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 500, in eval_master
    pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs
  File "/usr/lib/python2.7/dist-packages/salt/transport/client.py", line 154, in factory
    return salt.transport.zeromq.AsyncZeroMQPubChannel(opts, **kwargs
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 253, in __init__
    self.auth = salt.crypt.AsyncAuth(self.opts, io_loop=self.io_loop
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 324, in __new__
    new_auth.__singleton_init__(opts, io_loop=io_loop
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 362, in __singleton_init__
    self.get_keys
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 576, in get_keys
    salt.utils.verify.check_path_traversal(self.opts['pki_dir'], user
  File "/usr/lib/python2.7/dist-packages/salt/utils/verify.py", line 400, in check_path_traversal
    raise SaltClientError(msg
SaltClientError: Could not access /etc/salt/pki/minion. Try running as user root
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'master.local', 'tcp://127.0.0.1:4506', 'aes
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'master.local', 'tcp://127.0.0.1:4506
[DEBUG   ] No message_client attr for AsyncZeroMQReqChannel found. Not destroying sockets
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7f9bcfe01b50>> ignored
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7f9bcfe01b50>> ignored
```
If the `/etc/salt/pki/minion` directory does not exist, it will do the same thing but the SaltClientError message will change to saying "SaltClientError: Could not access /etc/salt/pki. Path does not exist.". I would also say that it is related to #27063 as well, although that issue is talking about the directory not existing, rather than Salt not having the permissions needed to read the directory.

If I am reading this output correctly, though, it fails to read the minion keys and then immediately after that tries to initialise an authentication with the salt master again. How could that possibly work if it can't read it's own minion keys?

**Comment by alangarf** [comment]
@ssgward yes it looks to be the same issue. Although I think at it's base is something to do with the comments in #29701. In #29701 a bandaid was placed on a wound to fix the issue in the packaging by making sure the `pki` dir is there. However it now appears the original issue in that ticket that wasn't resolved has presented itself again in another form.

**Comment by gladiatr72** [comment]
Where this hits me is on the salt-master--the minion (on the salt-master host) is running as root--

`/etc/salt/pki/minion` is owned by root--if salt-call is inadvertently run from the salt user (`_salt`), it cannot access the directory that is properly present with proper permissions. 

```
total 12
drwx------ 10 _salt _salt 4096 Feb  9 08:50 master
drwx------  2 root  root  4096 Dec 18 17:53 minion
```

**Comment by alangarf** [comment]
This is so annoying.

```
[21:26:50]-> salt-call event.send route53/update
[WARNING ] Failed to open log file, do you have permission to write to /var/log/salt/minion?
[WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate.
[ERROR   ] Future exception was never retrieved: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result)
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 500, in eval_master
    pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/transport/client.py", line 154, in factory
    return salt.transport.zeromq.AsyncZeroMQPubChannel(opts, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 253, in __init__
    self.auth = salt.crypt.AsyncAuth(self.opts, io_loop=self.io_loop)
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 324, in __new__
    new_auth.__singleton_init__(opts, io_loop=io_loop)
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 362, in __singleton_init__
    self.get_keys()
  File "/usr/lib/python2.7/dist-packages/salt/crypt.py", line 576, in get_keys
    salt.utils.verify.check_path_traversal(self.opts['pki_dir'], user)
  File "/usr/lib/python2.7/dist-packages/salt/utils/verify.py", line 400, in check_path_traversal
    raise SaltClientError(msg)
SaltClientError: Could not access /etc/salt/pki/minion. Try running as user root.
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7fd1e7b3bed0>> ignored
Exception ValueError: 'I/O operation on closed file' in <bound method SyncWrapper.__del__ of <salt.utils.async.SyncWrapper object at 0x7fd1e7b3bed0>> ignored
[LOOP LOOP LOOP]
^Z
[1]+  Stopped                 salt-call event.send learnosity/route53/updateA

[21:28:44]-> kill %1
```

Even just trying to send an event causes this loop if you forget to use `sudo`, and you can't interrupt it, you can only background the task and kill it. Although not before it's output a thousand plus lines of 'I/O operation on closed file' errors. :-1:

**Comment by ndobbs** [comment]
I too am seeing this issue, its effectively broken the custom 'post receive hook' I have with gitlab that calls for gitfs to automatically get updated during a git push operation.

**Comment by cachedout** [comment]
This has been fixed in the most recent release of 2015.8

```
mp@silver ...salt/salt/cli % salt-call test.ping                                                  (git)-[2015.8] 
14:26:43,942 [salt.log.setup                           ][WARNING ] Failed to open log file, do you have permission to write to /var/log/salt/minion?
14:26:44,334 [salt.loaded.int.grains.core              ][WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate.
Could not access /etc/salt/pki. Try running as user root.
```
